Cms/ProfileBioEdit.ascx.cs

```diff
@@ -0,0 +1,2394 @@
+                                        phoneNumber = new PhoneNumber { NumberTypeValueId = phoneNumberTypeId };
+                                        person.PhoneNumbers.Add( phoneNumber );
+                                    }
+
+                                    phoneNumber.CountryCode = PhoneNumber.CleanNumber( pnbPhone.CountryCode );
+                                    phoneNumber.Number = PhoneNumber.CleanNumber( pnbPhone.Number );
+
+                                    // Only allow one number to have SMS selected
+                                    if ( smsSelected )
+                                    {
+                                        phoneNumber.IsMessagingEnabled = false;
+                                    }
+                                    else
+                                    {
+                                        phoneNumber.IsMessagingEnabled = cbSms.Checked;
+                                        smsSelected = cbSms.Checked;
+                                    }
+
+                                    phoneNumberTypeIds.Add( phoneNumberTypeId );
+                                }
+                            }
+                        }
+                    }
+
+                    var phoneNumberService = new PhoneNumberService( rockContext );
+
+                    // Remove any duplicate numbers
+                    var hasDuplicate = person.PhoneNumbers.GroupBy( pn => pn.Number ).Where( g => g.Count() > 1 ).Any();
+
+                    if ( hasDuplicate )
+                    {
+                        var listOfValidNumbers = person.PhoneNumbers
+                            .OrderBy( o => o.NumberTypeValueId )
+                            .GroupBy( pn => pn.Number )
+                            .Select( y => y.First() )
+                            .ToList();
+                        var removedNumbers = person.PhoneNumbers.Except( listOfValidNumbers ).ToList();
+                        phoneNumberService.DeleteRange( removedNumbers );
+                        person.PhoneNumbers = listOfValidNumbers;
+                    }
+                }
+
+                if ( ebEmail != null )
+                {
+                    person.Email = ebEmail.Text.Trim();
+                }
+                if ( rblEmailPreference != null )
+                {
+                    person.EmailPreference = rblEmailPreference.SelectedValue.ConvertToEnum<EmailPreference>();
+                }
+
+                if ( rblCommunicationPreference != null )
+                {
+                    var communicationPreference = rblCommunicationPreference.SelectedValueAsEnum<CommunicationType>();
+
+                    if ( rblCommunicationPreference.Visible && selectedPhoneTypes.Any() && communicationPreference == CommunicationType.SMS && nbCommunicationPreferenceWarning != null )
+                    {
+                        if ( !person.PhoneNumbers.Any( a => a.IsMessagingEnabled ) )
```

Comment:
Test scenario where this code get's excercised, and verify it behaves as expected desired. Does it actually show the message, or does it kill the entire family save?

Reply:
it does what is expected now. If phone is showing you must have an SMS enabled phone number to set your communication preference to SMS, otherwise it will fail.